test/model/autocomplete_test.dart

```diff
@@ -349,4 +348,127 @@ void main() {
+
+      test('has DMs with userA and userB, latest with userB -> prioritizes userB', () async {
+        await prepare(dmConversations: [
+          RecentDmConversation(userIds: [idB],      maxMessageId: 200),
+          RecentDmConversation(userIds: [idA, idB], maxMessageId: 100),
+        ]);
+        check(compareAB()).isGreaterThan(0);
+      });
+
+      test('has DMs with userA and userB, equally recent -> prioritizes neither', () async {
+        await prepare(dmConversations: [
+          RecentDmConversation(userIds: [idA, idB], maxMessageId: 100),
+        ]);
+        check(compareAB()).equals(0);
+      });
+
+      test('has DMs with userA but not userB -> prioritizes userA', () async {
+        await prepare(dmConversations: [
+          RecentDmConversation(userIds: [idA], maxMessageId: 100),
+        ]);
+        check(compareAB()).isNegative();
+      });
+
+      test('has DMs with userB but not userA -> prioritizes userB', () async {
+        await prepare(dmConversations: [
+          RecentDmConversation(userIds: [idB], maxMessageId: 100),
+        ]);
+        check(compareAB()).isGreaterThan(0);
+      });
+
+      test('doesn\'t have DMs with userA or userB -> prioritizes neither', () async {
+        await prepare(dmConversations: []);
+        check(compareAB()).equals(0);
+      });
+    });
+
+    group('autocomplete suggests relevant users in the following order: '
+        '1. Users most recent in the DM conversations', () {
+      Future<void> checkResultsIn(Narrow narrow, {required List<int> expected}) async {
+        final users = [
+          eg.user(userId: 0),
+          eg.user(userId: 1),
+          eg.user(userId: 2),
+          eg.user(userId: 3),
+          eg.user(userId: 4),
+        ];
+
+        final dmConversations = [
+          RecentDmConversation(userIds: [3],    maxMessageId: 300),
+          RecentDmConversation(userIds: [0],    maxMessageId: 200),
+          RecentDmConversation(userIds: [0, 1], maxMessageId: 100),
+        ];
+
+        await prepare(users: users, dmConversations: dmConversations);
+        final view = MentionAutocompleteView.init(store: store, narrow: narrow);
+
+        bool done = false;
+        view.addListener(() { done = true; });
+        view.query = MentionAutocompleteQuery('');
+        await Future(() {});
+        check(done).isTrue();
+        final results = view.results
+          .map((e) => (e as UserMentionAutocompleteResult).userId)
+          .toList();
+        check(results).deepEquals(expected);
+      }
+
+      test('StreamNarrow', () async {
+        await checkResultsIn(const StreamNarrow(1), expected: [3, 0, 1, 2, 4]);
+      });
+
+      test('TopicNarrow', () async {
+        await checkResultsIn(const TopicNarrow(1, 'topic'), expected: [3, 0, 1, 2, 4]);
+      });
+
+      test('DmNarrow', () async {
+        await checkResultsIn(
+          DmNarrow(allRecipientIds: [eg.selfUser.userId],
+            selfUserId: eg.selfUser.userId),
```

Comment:
nit: can tighten a bit:
```suggestion
          DmNarrow.withUser(eg.selfUser.userId, selfUserId: eg.selfUser.userId),
```